src/main/java/com/google/sps/servlets/RegistrationServlet.java

```diff
@@ -26,23 +34,61 @@
 /** Servlet that handles user's registration info */
 @WebServlet("/registration")
 public class RegistrationServlet extends HttpServlet {
+  
+  private UserService userService = UserServiceFactory.getUserService();
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String role = request.getParameter("role");
-    String name = request.getParameter("name");
-    String email = request.getParameter("email");
+    String role = getParameter(request, "role").orElse(null);
+    String firstName = getParameter(request, "first-name")
+            .orElseThrow(() -> new IllegalArgumentException("Must fill out first name"));
+    String lastName = getParameter(request, "last-name")
+            .orElseThrow(() -> new IllegalArgumentException("Must fill out last name"));
+    String fullName = firstName + " " + lastName;
+    
+    String email = userService.getCurrentUser().getEmail();
+    String userId = userService.getCurrentUser().getUserId();
+    
     // Make list of selected topics, remove unchecked topics
-    List<String> topics = Arrays.asList(request.getParameter("math"), request.getParameter("english"), request.getParameter("other"));
-    topics = topics
-            .stream()
-            .filter(t -> t!= null)
-            .collect(Collectors.toList());
+    List<Optional<String>> topics = new ArrayList<Optional<String>>();
+    topics.add(getParameter(request, "math"));
+    topics.add(getParameter(request, "english"));
+    topics.add(getParameter(request, "other"));
+    List<String> topicsToStr = topics
+                                .stream()
+                                .filter(t -> t.isPresent())
+                                .map(t -> t.get())
+                                .collect(Collectors.toList());
 
-    response.setContentType("text/html;");
-    response.getWriter().println("<p>Student or tutor? " + role + "</p>");
-    response.getWriter().println("<p>Name: " + name + "</p>");
-    response.getWriter().println("<p>Email: " + email + "</p>");
-    response.getWriter().println("<p>Subject(s): " + topics + "</p>");
+    // Make entity for user with all registration info
+    Entity userEntity = new Entity("User");
```

Comment:
Can you put this line inside of `createEntityAndPutInDatastore`?

Reply:
I left it outside so I could use an entity as a parameter for `createEntityAndPutInDatastore` to make tests easier. Should I change the function name to make it less confusing?

Reply:
No, I think this works too! I missed the "used for testing" comment for `createEntityAndPutInDatastore`.